test/test_componentdefinition.py

```diff
@@ -240,5 +239,332 @@ def test_sequences_validation(self):
+                         CDS.identity, terminator.identity])
+
+    def test_compile_sequence(self):
+        doc = sbol2.Document()
+        sbol2.Config.setOption('sbol_typed_uris', True)
+        gene = sbol2.ComponentDefinition("BB0001")
+        promoter = sbol2.ComponentDefinition("R0010")
+        CDS = sbol2.ComponentDefinition("B0032")
+        RBS = sbol2.ComponentDefinition("E0040")
+        terminator = sbol2.ComponentDefinition("B0012")
+        scar = sbol2.ComponentDefinition('scar')
+
+        promoter.sequence = sbol2.Sequence('R0010')
+        RBS.sequence = sbol2.Sequence('B0032')
+        CDS.sequence = sbol2.Sequence('E0040')
+        terminator.sequence = sbol2.Sequence('B0012')
+        scar.sequence = sbol2.Sequence('scar')
+
+        promoter.sequence.elements = 'aaa'
+        RBS.sequence.elements = 'aaa'
+        CDS.sequence.elements = 'aaa'
+        terminator.sequence.elements = 'aaa'
+        scar.sequence.elements = 'ttt'
+
+        doc.addComponentDefinition(gene)
+        gene.assemblePrimaryStructure([promoter, scar, RBS, scar, CDS, scar,
+                                      terminator])
+        target_seq = gene.compile()
+
+        self.assertEqual(target_seq, 'aaatttaaatttaaatttaaa')
+        self.assertEqual(target_seq, gene.sequence.elements)
+
+    def test_recursive_compile(self):
+        doc = sbol2.Document()
+        cd1 = sbol2.ComponentDefinition('cd1')
+        cd2 = sbol2.ComponentDefinition('cd2')
+        cd3 = sbol2.ComponentDefinition('cd3')
+        cd4 = sbol2.ComponentDefinition('cd4')
+        cd5 = sbol2.ComponentDefinition('cd5')
+        cd1.sequence = sbol2.Sequence('cd1')
+        cd2.sequence = sbol2.Sequence('cd2')
+        cd3.sequence = sbol2.Sequence('cd3')
+        cd4.sequence = sbol2.Sequence('cd4')
+        cd5.sequence = sbol2.Sequence('cd5')
+        cd1.sequence.elements = 'tt'
+        cd2.sequence.elements = 'gg'
+        cd3.sequence.elements = 'n'
+        cd4.sequence.elements = 'aa'
+        cd5.sequence.elements = 'n'
+        doc.addComponentDefinition([cd1, cd2, cd3, cd4, cd5])
+        cd3.assemblePrimaryStructure([cd1, cd2])
+        cd5.assemblePrimaryStructure([cd4, cd3])
+        cd5.compile()
+        self.assertEquals(cd3.sequence.elements, 'ttgg')
+        self.assertEquals(cd5.sequence.elements, 'aattgg')
+        r1 = cd3.sequenceAnnotations['cd1_annotation_0'].\
+            locations['cd1_annotation_0_range']
+        r2 = cd3.sequenceAnnotations['cd2_annotation_0'].\
+            locations['cd2_annotation_0_range']
+        r4 = cd5.sequenceAnnotations['cd4_annotation_0'].\
+            locations['cd4_annotation_0_range']
+        self.assertEquals(r1.start, 3)
+        self.assertEquals(r1.end, 4)
+        self.assertEquals(r2.start, 5)
+        self.assertEquals(r2.end, 6)
+        self.assertEquals(r4.start, 1)
+        self.assertEquals(r4.end, 2)
+
+    @unittest.expectedFailure
```

Comment:
Why so many `expectedFailure` tests? Are these for the next wave of changes?

Reply:
Yes.  These were tests ported over from libSBOL but rely on some functionality that is still not implemented.  I think the next step would be investigate each test and file an issue.  I know at least one test failure is related to #55